File: tools/aconfig/aconfig_storage_write_api/tests/storage_write_api_test.cpp
/*
 * Copyright (C) 2024 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <string>
#include <vector>
#include <cstdio>

#include <sys/stat.h>
#include "aconfig_storage/aconfig_storage_read_api.hpp"
#include "aconfig_storage/aconfig_storage_write_api.hpp"
#include <gtest/gtest.h>
#include <android-base/file.h>
#include <android-base/result.h>

#include "rust/cxx.h"
#include "aconfig_storage/lib.rs.h"

using namespace android::base;

namespace api = aconfig_storage;
namespace private_api = aconfig_storage::private_internal_api;

class AconfigStorageTest : public ::testing::Test {
 protected:
  Result<std::string> copy_to_rw_temp_file(std::string const& source_file) {
    auto temp_file = std::string(std::tmpnam(nullptr));
    auto content = std::string();
    if (!ReadFileToString(source_file, &content)) {
      return Error() << "failed to read file: " << source_file;
    }
    if (!WriteStringToFile(content, temp_file)) {
      return Error() << "failed to copy file: " << source_file;
    }
    if (chmod(temp_file.c_str(),
              S_IRUSR | S_IRGRP | S_IROTH | S_IWUSR | S_IWGRP | S_IWOTH) == -1) {
      return Error() << "failed to chmod";
    }
    return temp_file;
  }

  void SetUp() override {
    auto const test_dir = android::base::GetExecutableDirectory();
    flag_val = *copy_to_rw_temp_file(test_dir + "/flag.val");
    flag_info = *copy_to_rw_temp_file(test_dir + "/flag.info");
  }

  void TearDown() override {
    std::remove(flag_val.c_str());
    std::remove(flag_info.c_str());
  }

  std::string flag_val;
  std::string flag_info;
};

/// Negative test to lock down the error when mapping a non writeable storage file
TEST_F(AconfigStorageTest, test_non_writable_storage_file_mapping) {
  ASSERT_TRUE(chmod(flag_val.c_str(), S_IRUSR | S_IRGRP | S_IROTH) != -1);
  auto mapped_file_result = api::map_mutable_storage_file(flag_val);
  ASSERT_FALSE(mapped_file_result.ok());
  auto it = mapped_file_result.error().message().find("cannot map nonwriteable file");
  ASSERT_TRUE(it != std::string::npos) << mapped_file_result.error().message();
}

/// Test to lock down storage flag value update api
TEST_F(AconfigStorageTest, test_boolean_flag_value_update) {
  auto mapped_file_result = api::map_mutable_storage_file(flag_val);
  ASSERT_TRUE(mapped_file_result.ok());
  auto mapped_file = std::unique_ptr<api::MutableMappedStorageFile>(*mapped_file_result);

  for (int offset = 0; offset < 8; ++offset) {
    auto update_result = api::set_boolean_flag_value(*mapped_file, offset, true);
    ASSERT_TRUE(update_result.ok());
    auto value = api::get_boolean_flag_value(*mapped_file, offset);
    ASSERT_TRUE(value.ok());
    ASSERT_TRUE(*value);
  }

  // load the file on disk and check has been updated
  std::ifstream file(flag_val, std::ios::binary | std::ios::ate);
  std::streamsize size = file.tellg();
  file.seekg(0, std::ios::beg);

  std::vector<uint8_t> buffer(size);
  file.read(reinterpret_cast<char *>(buffer.data()), size);

  auto content = rust::Slice<const uint8_t>(
      buffer.data(), mapped_file->file_size);

  for (int offset = 0; offset < 8; ++offset) {
    auto value_cxx = get_boolean_flag_value_cxx(content, offset);
    ASSERT_TRUE(value_cxx.query_success);
    ASSERT_TRUE(value_cxx.flag_value);
  }
}

/// Negative test to lock down the error when querying flag value out of range
TEST_F(AconfigStorageTest, test_invalid_boolean_flag_value_update) {
  auto mapped_file_result = api::map_mutable_storage_file(flag_val);
  ASSERT_TRUE(mapped_file_result.ok());
  auto mapped_file = std::unique_ptr<api::MutableMappedStorageFile>(*mapped_file_result);
  auto update_result = api::set_boolean_flag_value(*mapped_file, 8, true);
  ASSERT_FALSE(update_result.ok());
  ASSERT_EQ(update_result.error().message(),
            std::string("InvalidStorageFileOffset(Flag value offset goes beyond the end of the file.)"));
}

/// Test to lock down storage flag has server override update api
TEST_F(AconfigStorageTest, test_flag_has_server_override_update) {
  auto mapped_file_result = api::map_mutable_storage_file(flag_info);
  ASSERT_TRUE(mapped_file_result.ok());
  auto mapped_file = std::unique_ptr<api::MutableMappedStorageFile>(*mapped_file_result);

  for (int offset = 0; offset < 8; ++offset) {
    auto update_result = api::set_flag_has_server_override(
        *mapped_file, api::FlagValueType::Boolean, offset, true);
    ASSERT_TRUE(update_result.ok()) << update_result.error();
    auto attribute = api::get_flag_attribute(
        *mapped_file, api::FlagValueType::Boolean, offset);
    ASSERT_TRUE(attribute.ok());
    ASSERT_TRUE(*attribute & api::FlagInfoBit::HasServerOverride);
  }

  // load the file on disk and check has been updated
  std::ifstream file(flag_info, std::ios::binary | std::ios::ate);
  std::streamsize size = file.tellg();
  file.seekg(0, std::ios::beg);

  std::vector<uint8_t> buffer(size);
  file.read(reinterpret_cast<char *>(buffer.data()), size);

  auto content = rust::Slice<const uint8_t>(
      buffer.data(), mapped_file->file_size);

  for (int offset = 0; offset < 8; ++offset) {
    auto attribute = get_flag_attribute_cxx(content, api::FlagValueType::Boolean, offset);
    ASSERT_TRUE(attribute.query_success);
    ASSERT_TRUE(attribute.flag_attribute & api::FlagInfoBit::HasServerOverride);
  }

  for (int offset = 0; offset < 8; ++offset) {
    auto update_result = api::set_flag_has_server_override(
        *mapped_file, api::FlagValueType::Boolean, offset, false);
    ASSERT_TRUE(update_result.ok());
    auto attribute = api::get_flag_attribute(
        *mapped_file, api::FlagValueType::Boolean, offset);
    ASSERT_TRUE(attribute.ok());
    ASSERT_FALSE(*attribute & api::FlagInfoBit::HasServerOverride);
  }

  std::ifstream file2(flag_info, std::ios::binary);
  buffer.clear();
  file2.read(reinterpret_cast<char *>(buffer.data()), size);
  for (int offset = 0; offset < 8; ++offset) {
    auto attribute = get_flag_attribute_cxx(content, api::FlagValueType::Boolean, offset);
    ASSERT_TRUE(attribute.query_success);
    ASSERT_FALSE(attribute.flag_attribute & api::FlagInfoBit::HasServerOverride);
  }
}

/// Test to lock down storage flag has local override update api
TEST_F(AconfigStorageTest, test_flag_has_local_override_update) {
  auto mapped_file_result = api::map_mutable_storage_file(flag_info);
  ASSERT_TRUE(mapped_file_result.ok());
  auto mapped_file = std::unique_ptr<api::MutableMappedStorageFile>(*mapped_file_result);

  for (int offset = 0; offset < 8; ++offset) {
    auto update_result = api::set_flag_has_local_override(
        *mapped_file, api::FlagValueType::Boolean, offset, true);
    ASSERT_TRUE(update_result.ok());
    auto attribute = api::get_flag_attribute(
        *mapped_file, api::FlagValueType::Boolean, offset);
    ASSERT_TRUE(attribute.ok());
    ASSERT_TRUE(*attribute & api::FlagInfoBit::HasLocalOverride);
  }

  // load the file on disk and check has been updated
  std::ifstream file(flag_info, std::ios::binary | std::ios::ate);
  std::streamsize size = file.tellg();
  file.seekg(0, std::ios::beg);

  std::vector<uint8_t> buffer(size);
  file.read(reinterpret_cast<char *>(buffer.data()), size);

  auto content = rust::Slice<const uint8_t>(
      buffer.data(), mapped_file->file_size);

  for (int offset = 0; offset < 8; ++offset) {
    auto attribute = get_flag_attribute_cxx(content, api::FlagValueType::Boolean, offset);
    ASSERT_TRUE(attribute.query_success);
    ASSERT_TRUE(attribute.flag_attribute & api::FlagInfoBit::HasLocalOverride);
  }

  for (int offset = 0; offset < 8; ++offset) {
    auto update_result = api::set_flag_has_local_override(
        *mapped_file, api::FlagValueType::Boolean, offset, false);
    ASSERT_TRUE(update_result.ok());
    auto attribute = api::get_flag_attribute(
        *mapped_file, api::FlagValueType::Boolean, offset);
    ASSERT_TRUE(attribute.ok());
    ASSERT_FALSE(*attribute & api::FlagInfoBit::HasLocalOverride);
  }

  std::ifstream file2(flag_info, std::ios::binary);
  buffer.clear();
  file2.read(reinterpret_cast<char *>(buffer.data()), size);
  for (int offset = 0; offset < 8; ++offset) {
    auto attribute = get_flag_attribute_cxx(content, api::FlagValueType::Boolean, offset);
    ASSERT_TRUE(attribute.query_success);
    ASSERT_FALSE(attribute.flag_attribute & api::FlagInfoBit::HasLocalOverride);
  }
}

File: tools/aconfig/aconfig_storage_write_api/tests/storage_write_api_test.rs
#[cfg(not(feature = "cargo"))]
mod aconfig_storage_write_api_test {
    use aconfig_storage_file::{FlagInfoBit, FlagValueType};
    use aconfig_storage_read_api::flag_info_query::find_flag_attribute;
    use aconfig_storage_read_api::flag_value_query::find_boolean_flag_value;
    use aconfig_storage_write_api::{
        map_mutable_storage_file, set_boolean_flag_value, set_flag_has_local_override,
        set_flag_has_server_override,
    };

    use std::fs::{self, File};
    use std::io::Read;
    use tempfile::NamedTempFile;

    /// Create temp file copy
    fn copy_to_temp_rw_file(source_file: &str) -> NamedTempFile {
        let file = NamedTempFile::new().unwrap();
        fs::copy(source_file, file.path()).unwrap();
        file
    }

    /// Get boolean flag value from offset
    fn get_boolean_flag_value_at_offset(file: &str, offset: u32) -> bool {
        let mut f = File::open(file).unwrap();
        let mut bytes = Vec::new();
        f.read_to_end(&mut bytes).unwrap();
        find_boolean_flag_value(&bytes, offset).unwrap()
    }

    /// Get flag attribute at offset
    fn get_flag_attribute_at_offset(file: &str, value_type: FlagValueType, offset: u32) -> u8 {
        let mut f = File::open(file).unwrap();
        let mut bytes = Vec::new();
        f.read_to_end(&mut bytes).unwrap();
        find_flag_attribute(&bytes, value_type, offset).unwrap()
    }

    #[test]
    /// Test to lock down flag value update api
    fn test_boolean_flag_value_update() {
        let flag_value_file = copy_to_temp_rw_file("./flag.val");
        let flag_value_path = flag_value_file.path().display().to_string();

        // SAFETY:
        // The safety here is ensured as only this single threaded test process will
        // write to this file
        let mut file = unsafe { map_mutable_storage_file(&flag_value_path).unwrap() };
        for i in 0..8 {
            set_boolean_flag_value(&mut file, i, true).unwrap();
            let value = get_boolean_flag_value_at_offset(&flag_value_path, i);
            assert!(value);

            set_boolean_flag_value(&mut file, i, false).unwrap();
            let value = get_boolean_flag_value_at_offset(&flag_value_path, i);
            assert!(!value);
        }
    }

    #[test]
    /// Test to lock down flag has server override update api
    fn test_set_flag_has_server_override() {
        let flag_info_file = copy_to_temp_rw_file("./flag.info");
        let flag_info_path = flag_info_file.path().display().to_string();

        // SAFETY:
        // The safety here is ensured as only this single threaded test process will
        // write to this file
        let mut file = unsafe { map_mutable_storage_file(&flag_info_path).unwrap() };
        for i in 0..8 {
            set_flag_has_server_override(&mut file, FlagValueType::Boolean, i, true).unwrap();
            let attribute =
                get_flag_attribute_at_offset(&flag_info_path, FlagValueType::Boolean, i);
            assert!((attribute & (FlagInfoBit::HasServerOverride as u8)) != 0);
            set_flag_has_server_override(&mut file, FlagValueType::Boolean, i, false).unwrap();
            let attribute =
                get_flag_attribute_at_offset(&flag_info_path, FlagValueType::Boolean, i);
            assert!((attribute & (FlagInfoBit::HasServerOverride as u8)) == 0);
        }
    }

    #[test]
    /// Test to lock down flag has local override update api
    fn test_set_flag_has_local_override() {
        let flag_info_file = copy_to_temp_rw_file("./flag.info");
        let flag_info_path = flag_info_file.path().display().to_string();

        // SAFETY:
        // The safety here is ensured as only this single threaded test process will
        // write to this file
        let mut file = unsafe { map_mutable_storage_file(&flag_info_path).unwrap() };
        for i in 0..8 {
            set_flag_has_local_override(&mut file, FlagValueType::Boolean, i, true).unwrap();
            let attribute =
                get_flag_attribute_at_offset(&flag_info_path, FlagValueType::Boolean, i);
            assert!((attribute & (FlagInfoBit::HasLocalOverride as u8)) != 0);
            set_flag_has_local_override(&mut file, FlagValueType::Boolean, i, false).unwrap();
            let attribute =
                get_flag_attribute_at_offset(&flag_info_path, FlagValueType::Boolean, i);
            assert!((attribute & (FlagInfoBit::HasLocalOverride as u8)) == 0);
        }
    }
}
